FreeRTOS-Plus/Source/Application-Protocols/network_transport/freertos_plus_tcp/using_wolfSSL/using_wolfSSL.c

```diff
@@ -0,0 +1,409 @@
+*  @param[in] buf Buffer for received data
+*  @param[in] sz  Size to receive
+*  @param[in] context Socket to be received from 
+*  
+*  @return received size( > 0 ), #WOLFSSL_CBIO_ERR_CONN_CLOSE, #WOLFSSL_CBIO_ERR_WANT_READ. 
+*/
+static int wolfSSL_IORecvGlue(WOLFSSL* ssl, char* buf, int sz, void* context);
+
+
+
+
+
+/*-----------------------------------------------------------*/
+static int wolfSSL_IORecvGlue(WOLFSSL* ssl, char* buf, int sz, void* context)
+{
+    ssl;    /* to prevent unused warning*/
+
+    Socket_t xSocket = (Socket_t)context;
+
+    BaseType_t read;
+  
+    read = FreeRTOS_recv(xSocket, (void*)buf, (size_t)sz, 0);
+
+
+    if ((read == 0 ) ||
+        (read == -pdFREERTOS_ERRNO_EWOULDBLOCK))
+    {
+        read = WOLFSSL_CBIO_ERR_WANT_READ;
+    }
+    else if (read == -pdFREERTOS_ERRNO_ENOTCONN)
+    {
+        read = WOLFSSL_CBIO_ERR_CONN_CLOSE;
+    }
+    else
+    {
+        /* do nothing */
+    }
+    return (int)read;
+}
+/*-----------------------------------------------------------*/
+
+static int wolfSSL_IOSendGlue(WOLFSSL* ssl, char* buf, int sz, void* context)
+{
+    ssl;    /* to prevent unused warning*/
+
+    Socket_t xSocket = (Socket_t)context;
+
+    BaseType_t sent = FreeRTOS_send(xSocket, (void*)buf, (size_t)sz, 0);
+
+    if ( sent == -pdFREERTOS_ERRNO_EWOULDBLOCK )
+    {
+        sent = WOLFSSL_CBIO_ERR_WANT_WRITE;
+    }
+    else if ( sent == -pdFREERTOS_ERRNO_ENOTCONN )
+    {
+        sent = WOLFSSL_CBIO_ERR_CONN_CLOSE;
+    }
+    else
+    {
+        /* do nothing */
+    }
+    return (int)sent;
+}
+
+/*-----------------------------------------------------------*/
+static TlsTransportStatus_t initTLS(void)
+{
+    /* initialize wolfSSL */
+    wolfSSL_Init();
+
+#ifdef DEBUG_WOLFSSL
+    wolfSSL_Debugging_ON();
+#endif
+
+    return TLS_TRANSPORT_SUCCESS;
+}
+
+/*-----------------------------------------------------------*/
+
+static TlsTransportStatus_t tlsSetup( NetworkContext_t * pNetworkContext,
+                                      const char * pHostName,
+                                      const NetworkCredentials_t * pNetworkCredentials )
+{
+    TlsTransportStatus_t returnStatus = TLS_TRANSPORT_SUCCESS;
+    int iResult = SSL_SUCCESS;
+    /*char errString[80];*/
+
+    configASSERT( pNetworkContext != NULL );
+    configASSERT( pHostName != NULL );
+    configASSERT( pNetworkCredentials != NULL );
+    configASSERT( pNetworkCredentials->pRootCa != NULL );
+    configASSERT( pNetworkContext->tcpSocket != NULL );
+    
+    if (pNetworkContext->sslContext.ctx == NULL)
+    {
+        /* Attempt to create a context that uses the TLS 1.3 or 1.2 client protocol. */
+        pNetworkContext->sslContext.ctx = wolfSSL_CTX_new(wolfSSLv23_client_method_ex(NULL));
+    }
+
+    if (pNetworkContext->sslContext.ctx == NULL)
+    {
+        LogError(("Failed to create a wolfSSL_CTX"));
+        returnStatus = TLS_TRANSPORT_CONNECT_FAILURE;
+        return returnStatus;
```

Comment:
This should be modified slightly in order to have a single point of return

Reply:
Rewrote this function to have a single point of return.